### PR TITLE
[GSoC] Functions: Adds cdir parameter to handle series expansions on branch cuts

### LINKED ISSUE #7216: Limits involving branch cuts of elementary functions not handled
Branch cuts of sqrt() and log() are not handled properly.

If that implies that sqrt() and log() have more than this issue in the complex plane I don't know.
Moreover, if that implies that other elementary functions are having similar problems with branch cuts I have not tested.
```
In [1]: from sympy import *                                        
```
Exmpl. 1-------------------sqrt()---------------------------
1st, numerically
```
In [2]: N(sqrt(-1.+I*1e-10))                                                                                   
Out[2]: 5.0e-11 + 1.0*I

In [3]: N(sqrt(-1.-I*1e-10))                                                                                   
Out[3]: 5.0e-11 - 1.0*I             
```
Ok. sqrt-branch seems to lie on neg. Re-axis and is selected properly
2nd, same thing symbolically
```
In [5]: eps=symbols('eps')                                                                                     
In [6]: limit(sqrt(-1+I*eps),eps,0,'+')                                                                        
Out[6]: I

In [7]: limit(sqrt(-1+I*eps),eps,0,'-')                                                                        
Out[7]: I

In [8]: limit(sqrt(-1-I*eps),eps,0,'+')                                                                        
Out[8]: I

In [9]: limit(sqrt(-1-I*eps),eps,0,'-')                                                                        
Out[9]: I
```
Symbolic sqrt() returns 'I' irrespective of approach to branch cut. Wrong.

Exmpl. 2-------------log()--------------
1st, numerically
```
In [10]: log(-1I*1e-10)                                                                                       
Out[10]: 5.0e-21 - 3.14159265348979*I

In [11]: log(-1+I*1e-10)                                                                                       
Out[11]: 5.0e-21 + 3.14159265348979*I
```
Ok. log-branch seems to lie on neg. Re-axis and is selected properly
2nd, same thing symbolically
```
In [12]: limit(log(-1+I*eps),eps,0,'+')                                                                         
Out[12]: I*pi

In [13]: limit(log(-1+I*eps),eps,0,'-')                                                                         
Out[13]: I*pi

In [14]: limit(log(-1-I*eps),eps,0,'+')                                                                         
Out[14]: I*pi

In [15]: limit(log(-1-I*eps),eps,0,'-')                                                                         
Out[15]: I*pi
```
Symbolic log() returns 'I*pi' irrespective of approach to branch cut. Wrong.

Original issue for #7216: http://code.google.com/p/sympy/issues/detail?id=4117
Original author: https://code.google.com/u/112689286102358441542/

### LINKED ISSUE #19586: Attribute error in limit
`limit(x**(2**x*3**(-x)),x,oo)` raises `AttributeError: 'Pow' object has no attribute 'ceiling'`. Tested using Sympy 1.6
Everything was OK in 1.5 version.

### PR DESCRIPTION
**Pull request by sachin-4099**
Fixes: #7216
Fixes: #19586 

#### Brief description of what is fixed or changed

Adds `cdir` parameter to handle `series expansions` on `branch cuts`.
`cdir` stands for `complex direction`, and indicates the direction from which the expansion needs to be evaluated.

#### Other Comments
Regression Tests have been added.

#### Release Notes

* functions
  * Adds `cdir` parameter to handle `series expansions` on `branch cuts`

### CODE REVIEW

File: sympy/core/expr.py [hunk 1]
```diff
@@ -1020,6 +1020,17 @@ def conjugate(self):
         from sympy.functions.elementary.complexes import conjugate as c
         return c(self)
 
+    def dir(self, x, cdir):
+        minexp = S.Zero
+        arg = self
```
Comment (jksuom) [inline sympy/core/expr.py:1025]: It should be checked somewhere, maybe here, that `arg` is not 0 to begin with.

Comment (sachin-4099) [reply sympy/core/expr.py:1025]: So something like:
```
if arg.is_zero:
    return
```

Comment (jksuom) [reply sympy/core/expr.py:1025]: How about retuning 0 which is the default?

Comment (sachin-4099) [reply sympy/core/expr.py:1025]: should work fine.

File: sympy/core/mul.py [hunk 2]
```diff
@@ -1791,10 +1794,18 @@ def coeff_exp(term, x):
                     raise ValueError
 
             n0 = sum(t[1] for t in ords)
-            facs = [t.series(x, 0, ceiling(n-n0+m)).removeO() for t, m in ords]
+            facs = []
+            for t, m in ords:
+                n1 = ceiling(n - n0 + m)
+                s = t.nseries(x, 0, n1)
```
Comment (jksuom) [inline sympy/core/mul.py:1800]: `logx` and `cdir` should probably be added here.

File: sympy/core/power.py [hunk 3]
```diff
@@ -1575,7 +1578,10 @@ def mul(d1, d2):
             tk = mul(tk, gterms)
             k += S.One
 
-        inco, inex = coeff_exp(f**e, x)
+        if not e.is_integer and m.is_zero and f.is_real and f.is_negative and im((b - f).dir(x, cdir)) < 0:
```
Comment (jksuom) [inline sympy/core/power.py:1581]: This is a long(ish) line.

File: sympy/core/tests/test_power.py [hunk 4]
```diff
@@ -277,6 +277,14 @@ def test_pow_as_base_exp():
     assert Pow(1, 2, evaluate=False).as_base_exp() == (S.One, S(2))
 
 
+def test_nseries():
+    x = Symbol('x')
+    assert sqrt(I*x - 1)._eval_nseries(x, 4, None, 1) == I + x/2 + I*x**2/8 - x**3/16 + O(x**4)
+    assert sqrt(I*x - 1)._eval_nseries(x, 4, None, -1) == -I - x/2 - I*x**2/8 + x**3/16 + O(x**4)
+    assert cbrt(I*x - 1)._eval_nseries(x, 4, None, 1) == (-1)**(S(1)/3) - (-1)**(S(5)/6)*x/3 + (-1)**(S(1)/3)*x**2/9 + 5*(-1)**(S(5)/6)*x**3/81 + O(x**4)
```
Comment (jksuom) [inline sympy/core/tests/test_power.py:284]: These two long `cbrt` lines should be split.

File: sympy/core/power.py (hunk 3 above)
Comment (sachin-4099) [reply sympy/core/power.py:1581]: Will reduce its length using parenthesis.

**Comment by jksuom** [comment]
Thanks, I think this is ready to go.

**Pull request merged**